File: bindings/elixir/lib/proven/safe_probability.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

defmodule Proven.SafeProbability do
  @moduledoc """
  Safe probability operations via libproven FFI.

  Probability clamping and composition are performed by the
  Idris 2 verified core.
  """

  @doc """
  Create a probability value clamped to [0.0, 1.0].
  """
  @spec create(float()) :: float()
  def create(value) when is_float(value) do
    Proven.NIF.nif_probability_create(value)
  end

  @doc """
  Compute P(A and B) for independent events.
  """
  @spec prob_and(float(), float()) :: float()
  def prob_and(a, b) when is_float(a) and is_float(b) do
    Proven.NIF.nif_probability_and(a, b)
  end

  @doc """
  Compute P(A or B) for mutually exclusive events.
  """
  @spec prob_or(float(), float()) :: float()
  def prob_or(a, b) when is_float(a) and is_float(b) do
    Proven.NIF.nif_probability_or(a, b)
  end

  @doc """
  Compute P(not A) = 1 - P(A).
  """
  @spec complement(float()) :: float()
  def complement(p) when is_float(p) do
    Proven.NIF.nif_probability_not(p)
  end
end

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafeProbability.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

/**
 * Safe probability operations via libproven FFI.
 *
 * Provides probability creation, conjunction, disjunction, and negation.
 * Values are clamped to [0.0, 1.0] by the native library.
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No probability logic is reimplemented in Java.
 */
public final class SafeProbability {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    private SafeProbability() {}

    /**
     * Create a probability value, clamped to [0.0, 1.0].
     *
     * @param value the raw probability value
     * @return the clamped probability
     */
    public static double create(double value) {
        return LIB.proven_probability_create(value);
    }

    /**
     * Probability of independent events both occurring (P(A) * P(B)).
     *
     * @param a probability of event A
     * @param b probability of event B
     * @return the joint probability
     */
    public static double and(double a, double b) {
        return LIB.proven_probability_and(a, b);
    }

    /**
     * Probability of exactly one event occurring (exclusive or).
     *
     * @param a probability of event A
     * @param b probability of event B
     * @return the exclusive-or probability
     */
    public static double orExclusive(double a, double b) {
        return LIB.proven_probability_or_exclusive(a, b);
    }

    /**
     * Complement probability (1 - P).
     *
     * @param p the probability
     * @return the complement
     */
    public static double not(double p) {
        return LIB.proven_probability_not(p);
    }
}
